File: samples/csharp_dotnetcore/45.state-management/Bots/StateManagementBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples
{
    public class StateManagementBot : ActivityHandler
    {
        private BotState _conversationState;
        private BotState _userState;

        public StateManagementBot(ConversationState conversationState, UserState userState)
        {
            _conversationState = conversationState;
            _userState = userState;
        }

        public override async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
        {
            await base.OnTurnAsync(turnContext, cancellationToken);

            // Save any state changes that might have occured during the turn.
            await _conversationState.SaveChangesAsync(turnContext, false, cancellationToken);
            await _userState.SaveChangesAsync(turnContext, false, cancellationToken);
        }

        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            await turnContext.SendActivityAsync("Welcome to State Bot Sample. Type anything to get started.");
        }

        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            // Get the state properties from the turn context.

            var conversationStateAccessors =  _conversationState.CreateProperty<ConversationData>(nameof(ConversationData));
            var conversationData = await conversationStateAccessors.GetAsync(turnContext, () => new ConversationData());

            var userStateAccessors = _userState.CreateProperty<UserProfile>(nameof(UserProfile));
            var userProfile = await userStateAccessors.GetAsync(turnContext, () => new UserProfile());

            if (string.IsNullOrEmpty(userProfile.Name))
            {
                // First time around this is set to false, so we will prompt user for name.
                if (conversationData.PromptedUserForName)
                {
                    // Set the name to what the user provided.
                    userProfile.Name = turnContext.Activity.Text?.Trim();

                    // Acknowledge that we got their name.
                    await turnContext.SendActivityAsync($"Thanks {userProfile.Name}. To see conversation data, type anything.");

                    // Reset the flag to allow the bot to go though the cycle again.
                    conversationData.PromptedUserForName = false;
                }
                else
                {
                    // Prompt the user for their name.
                    await turnContext.SendActivityAsync($"What is your name?");

                    // Set the flag to true, so we don't prompt in the next turn.
                    conversationData.PromptedUserForName = true;
                }
            }
            else
            {
                // Add message details to the conversation data.
                // Convert saved Timestamp to local DateTimeOffset, then to string for display.
                var messageTimeOffset = (DateTimeOffset) turnContext.Activity.Timestamp;
                var localMessageTime = messageTimeOffset.ToLocalTime();
                conversationData.Timestamp = localMessageTime.ToString();
                conversationData.ChannelId = turnContext.Activity.ChannelId.ToString();

                // Display state data.
                await turnContext.SendActivityAsync($"{userProfile.Name} sent: {turnContext.Activity.Text}");
                await turnContext.SendActivityAsync($"Message received at: {conversationData.Timestamp}");
                await turnContext.SendActivityAsync($"Message received from: {conversationData.ChannelId}");
            }
        }
    }
}

File: samples/java_springboot/45.state-management/src/main/java/com/microsoft/bot/sample/statemanagement/StateManagementBot.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.statemanagement;

import com.codepoetics.protonpack.collectors.CompletableFutures;
import com.microsoft.bot.builder.ActivityHandler;
import com.microsoft.bot.builder.ConversationState;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.builder.StatePropertyAccessor;
import com.microsoft.bot.builder.TurnContext;
import com.microsoft.bot.builder.UserState;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.ChannelAccount;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * This class implements the functionality of the Bot.
 *
 * <p>
 * This is where application specific logic for interacting with the users would
 * be added. This class tracks the conversation state through POJO's saved in
 * {@link ConversationState} and {@link UserState}.
 * </p>
 *
 * @see ConversationData
 * @see UserProfile
 */
public class StateManagementBot extends ActivityHandler {
    private ConversationState conversationState;
    private UserState userState;

    @Autowired
    public StateManagementBot(ConversationState withConversationState, UserState withUserState) {
        conversationState = withConversationState;
        userState = withUserState;
    }

    /**
     * Normal onTurn processing, with saving of state after each turn.
     *
     * @param turnContext The context object for this turn. Provides information
     *                    about the incoming activity, and other data needed to
     *                    process the activity.
     * @return A future task.
     */
    @Override
    public CompletableFuture<Void> onTurn(TurnContext turnContext) {
        return super.onTurn(turnContext)
            .thenCompose(turnResult -> conversationState.saveChanges(turnContext))
            .thenCompose(saveResult -> userState.saveChanges(turnContext));
    }

    /**
     * Send a welcome message to new members.
     *
     * @param membersAdded A list of all the members added to the conversation, as
     *                     described by the conversation update activity.
     * @param turnContext  The context object for this turn.
     * @return A future task.
     */
    @Override
    protected CompletableFuture<Void> onMembersAdded(
        List<ChannelAccount> membersAdded,
        TurnContext turnContext
    ) {
        return membersAdded.stream()
            .filter(
                member -> !StringUtils
                    .equals(member.getId(), turnContext.getActivity().getRecipient().getId())
            )
            .map(
                channel -> turnContext
                    .sendActivity(
                        MessageFactory
                            .text("Welcome to State Bot Sample. Type anything to get started.")
                    )
            )
            .collect(CompletableFutures.toFutureList())
            .thenApply(resourceResponses -> null);
    }

    /**
     * This will prompt for a user name, after which it will send info about the
     * conversation. After sending information, the cycle restarts.
     *
     * @param turnContext The context object for this turn.
     * @return A future task.
     */
    @Override
    protected CompletableFuture<Void> onMessageActivity(TurnContext turnContext) {
        // Get state data from ConversationState.
        StatePropertyAccessor<ConversationData> dataAccessor =
            conversationState.createProperty("data");
        CompletableFuture<ConversationData> dataFuture =
            dataAccessor.get(turnContext, ConversationData::new);

        // Get profile from UserState.
        StatePropertyAccessor<UserProfile> profileAccessor = userState.createProperty("profile");
        CompletableFuture<UserProfile> profileFuture =
            profileAccessor.get(turnContext, UserProfile::new);

        return dataFuture.thenCombine(profileFuture, (conversationData, userProfile) -> {
            if (StringUtils.isEmpty(userProfile.getName())) {
                if (conversationData.getPromptedUserForName()) {
                    // Reset the flag to allow the bot to go though the cycle again.
                    conversationData.setPromptedUserForName(false);

                    // Set the name to what the user provided and reply.
                    userProfile.setName(turnContext.getActivity().getText());
                    return turnContext.sendActivity(
                        MessageFactory.text(
                            "Thanks " + userProfile.getName()
                                + ".  To see conversation data, type anything."
                        )
                    );
                } else {
                    conversationData.setPromptedUserForName(true);
                    return turnContext.sendActivity(MessageFactory.text("What is your name?"));
                }
            } else {
                // Set the flag to true, so we don't prompt in the next turn.
                conversationData.setPromptedUserForName(true);

                OffsetDateTime messageTimeOffset = turnContext.getActivity().getTimestamp();
                LocalDateTime localMessageTime = messageTimeOffset.toLocalDateTime();
                conversationData.setTimestamp(localMessageTime.toString());
                conversationData.setChannelId(turnContext.getActivity().getChannelId());

                List<Activity> sendToUser = new ArrayList<>();

                sendToUser.add(
                    MessageFactory.text(
                        userProfile.getName() + " sent: " + turnContext.getActivity().getText()
                    )
                );

                sendToUser.add(
                    MessageFactory.text(
                        userProfile.getName() + " message received at: "
                            + conversationData.getTimestamp()
                    )
                );

                sendToUser.add(
                    MessageFactory.text(
                        userProfile.getName() + " message received from: "
                            + conversationData.getChannelId()
                    )
                );

                return turnContext.sendActivities(sendToUser);
            }
        })
            // make the return value happy.
            .thenApply(resourceResponse -> null);
    }
}
